File: prog/osloware/reduc/src/reduc_winformat.f
C    ===========================================================
      SUBROUTINE reduce_format(  
     +                           rawdata,
     +                           outp,
     +                           messp,
     +                           final_egain,
     +                           writestat,
     +                           errstat )

C    ===========================================================
C     Purpose : Sort data into a matrix which is used by       C
C               the sorting application routine.               C
C               The event matrix is processed by the subr.     C
C               reduce_sort which returns whether the event    C
C               is to be stored or not.                        C
C                                                              C
C     INPUT  via parameter : rawdata : integer                 C
C     OUTPUT via parameter :                                   C
C                            errstat : integer=-1 buffer err   C
C                            writestat : integer=-1 write err  C
C                                                              C
C     OUTPUT via common    : COMMON /event/matrix(0:11,0:15)   C
C                                                              C
C     =========================================================C
      IMPLICIT NONE

      INCLUDE 'common_defs.decl'

      integer rawdata(0:32767)
      real final_egain(0:63) 		
      integer errstat
      integer writestat
      integer messp ,outp  
      integer writestatus
      real randval1, randval2, rand
      real evfrac

      integer OP			! "Pointer" used in local databuffer
      integer rawevent(0:1023)
      integer bufferlength 
      integer grec,gevents,events
      integer i,j,k,l,n
      integer evlen
      integer tel
      integer header_id,header_mask,temp,tpu
      integer xindex,yindex,xrow
      integer nextpointer,currentpointer,header,pattern
      logical good

C --  Window (gate values)
      integer alfa_low,alfa_high,h_low,h_high                                   
      common /window/alfa_low(0:63),alfa_high,h_low,h_high 

C --  Write to Exabyte tape
      integer write_exa
      external write_exa !$pragma C( write_exa )

C --  Get access to shared message box segment
      integer message_box(0:19)
      pointer (messp, message_box)	

C --  Local databuffer with accepted events
      integer OutBuf(0:32767)
      pointer (outp, OutBuf) ! Pointer to output buffer used by C-routine write_exa
      bufferlength = 32767

C     Bit masks
      header_id   = X'F000'                 ! Event header (Hex)
      header_mask = X'FF'        

C --  Initialize Variables
      gevents         = 0
      events          = 0
      errstat         = 0  
      writestat       = 0
      currentpointer  = 0
      nextpointer     = 0 
      good            = .FALSE.

C ======================================================================
      do i = 0,bufferlength                 ! Event loop

c         if(currentpointer.GT.300.AND.currentpointer.LT.315.AND.message_box(5).EQ.1)then ! Magnes test
c           WRITE(*,*)'word  rawdata reducdata'
c           do n=0,300
c              write(*,1111)n,rawdata(n),OutBuf(n)
c 1111         format(I4,Z8,Z8)
c           enddo
c         endif

         m(0,0) = 0                         ! Reset TPU id loc.
         m(3,0) = 0                         !         -"-
         m(6,0) = 0                         !         -"-
         m(9,0) = 0                         !         -"-

         currentpointer = nextpointer       ! Pointer to current
         header = rawdata(currentpointer)   ! Start-Of-Event

C --     No more data in buffer if header=0,return from subroutine
C --     If this should occur to early (partly filled buffer) it may
C --     be a serious problem. Mark such a buffer as bad buffer.
         if (header .EQ. 0 .AND. currentpointer .GT. 32500 ) goto 999

C --     Check header for start-of-event code (=Fxxx)
           temp = header .AND. header_id
           if (temp .ne. header_id) then         ! Error has occured
              errstat = -1
              goto 999                           ! Drop current buffer
           end if

C --     Save pointer to next event 
           evlen = header .AND. header_mask
           IF(evlen.GT.1024)THEN
              WRITE(*,*)'Warning, eventlength > 1024'
              errstat = -1
              GO TO 999
           ENDIF
           nextpointer = evlen + currentpointer


C --     Save the raw event
           do l= 0,evlen - 1
              rawevent(l) = rawdata(currentpointer + l)
           end do

           currentpointer = currentpointer + 1

C --  ------------------------------------------------------------------
      do while (currentpointer .LT. nextpointer) ! Sub-event loop

C --     Identify TPU
         tpu = ((rawdata(currentpointer)) .and. X'F')
         if      (tpu .EQ. X'A') then
            xindex = 0
         else if (tpu .EQ. X'B') then
            xindex = 3
         else if (tpu .EQ. X'C') then
            xindex = 6
         else if (tpu .EQ. X'D') then
            xindex = 9
         end if
         currentpointer = currentpointer + 1

         if (xindex .ne. 9) then                   ! No pattern for PUR
C --        Put the pattern into matrix
            pattern = rawdata(currentpointer)
            m(xindex,1) = pattern
            currentpointer = currentpointer + 1
         end if



C ---------------------------------------------------------------------
C       T P U 1 :
C 
C     Fill in data according to pattern.
C     For TPU1 only even-numbered pattern bits have two data words.
C     The odd-numbered are logical bits which may be used by the
C     sorting routine.

      if (xindex .eq. 0) then                      ! TPU1 is identified
         m(xindex,0) = 1

         do j = 0,14,2
             if (btest(pattern,j)) then
                 xrow=xindex+1
                 yindex=j
                 m(xrow,yindex) = rawdata(currentpointer)
                 currentpointer = currentpointer + 1
                 xrow=xindex+2
                 m(xrow,yindex) = rawdata(currentpointer)
                 currentpointer = currentpointer + 1
             end if
         end do

C ---------------------------------------------------------------------
C       T P U 2 :
C 
C     Fill in data according to pattern
C     For TPU all pattern bits have two data words, the first may
C     i.e. denote an energy while word two contains the time 

      else if (xindex .eq. 3) then                  ! TPU2 is identified
         m(xindex,0) = 1

         do l = 0,15
             if (btest(pattern,l)) then
                 xrow=xindex+1
                 yindex=l
                 m(xrow,yindex) = rawdata(currentpointer)
                 currentpointer = currentpointer + 1
                 xrow=xindex+2
                 m(xrow,yindex) = rawdata(currentpointer)
                 currentpointer = currentpointer + 1
             end if
         end do

C ---------------------------------------------------------------------
C       T P U 3 :
C 
C     Fill in data according to pattern
C     For TPU all pattern bits have two data words, the first may
C     i.e. denote an energy while word two contains the time 

      else if (xindex .EQ. 6) then                  ! TPU3 is identified
         m(xindex,0) = 1

         do n = 0,15
             if (btest(pattern,n)) then
                 xrow=xindex+1
                 yindex=n
                 m(xrow,yindex) = rawdata(currentpointer)
                 currentpointer = currentpointer + 1
                 xrow=xindex+2
                 m(xrow,yindex) = rawdata(currentpointer)
                 currentpointer = currentpointer + 1
             end if
         end do

C-----------------------------------------------------------------------
C       T P U 4 :
C
C     Virtual TPU
C     The Pile-Up pattern words are stored under "virtual TPU 4"
C     energy words 0 and 1

      else if (xindex .EQ. 9) then             ! PUR is identified

         m(xindex,0) = 1

         m(10,0)   = rawdata(currentpointer)   ! PUR(0)
         currentpointer = currentpointer + 1
         m(10,1)   = rawdata(currentpointer)   ! PUR(1)
         currentpointer = currentpointer + 1
      end if

C --  Fetch next subevent
      end do                                   ! End-Of-Subevent loop
C-----------------------------------------------------------------------

C --  Now, the event is formatted and ready to be sorted
      call reduce_sort( good,tel )             ! Gate on particles
 
               
C --  Subroutine "reduce_sort" return good = .TRUE. if
C --  the event has a particle within a window in the thickness
C --  spectrum.

      events = events + 1
      if (good) then                           ! Good event, save it
         gevents = gevents + 1                 ! No of good events
         evfrac = (gevents * 100) / events
         message_box(14) = INT(evfrac)         ! % events accepted

C --     Output Databuffer containing accepted events is full
         if ( OP .gt. bufferlength-evlen) then 
   
C --        Write databuffer to Target Device
            writestatus = write_exa( %VAL(outp), %VAL(messp) )

            if (writestatus .EQ. -1) then      ! Error in write, terminate session
               writestat = -1
               goto 999
            end if 
                        
            grec = grec + 1                    ! Increment "good" buffers
            OP = 0                             ! Reset local output buffer pointer
            DO n = 0, bufferlength             ! Zeroing local output buffer
               OutBuf(n) = 0
            ENDDO
         end if 

	   message_box(15) = grec                ! Number of accepted buffers


C --   Find a random number between -0.5 and +0.5
         randval1 = rand(0) - 0.5
         randval2 = rand(0) - 0.5


C  ***
C  ***   This code must be modified to accomodate 64 telescopes
C  ***
C --     Multiply E and DE in raw event by final gain. The event has one 
C --     and only one particle subevent. This subevent is the first in the event.
C --     Thus the DE-energy word will always be located in rawevent(3) and the
C --     E-energy word in rawevent(4). The rawevent energy words must be normalized
C --     to [0,2047] in advance to gain multiplication and restored to full length
C --     before saved. 
C --     1. Telescope group A, tel = [0,3]
         if (tel .GE. 0 .AND. tel .LE. 3) then
            rawevent(3) = ((rawevent(3) - (tel*2048) + randval1) * gainde(tel)) + shiftde(tel) +0.5 
            if ( rawevent(3) .GT. 2047) rawevent(3) = 0
            if ( rawevent(3) .LT.    0) rawevent(3) = 0

            rawevent(3) = rawevent(3) + (tel*2048)
            rawevent(4) = ((rawevent(4) - (tel*2048) + randval2) * final_egain(tel)) + shifte(tel) + 0.5
            if ( rawevent(4) .GT. 2047) rawevent(4) = 0
            if ( rawevent(4) .LT.    0) rawevent(4) = 0
            rawevent(4) = rawevent(4) + (tel*2048) 
         end if
C --     2. Telescope group B, tel = [4,7]
         if (tel .GE. 4 .AND. tel .LE. 7) then
            rawevent(3) = ((rawevent(3) - ((tel-4)*2048) + randval1) * gainde(tel)) + shiftde(tel) + 0.5 
            if ( rawevent(3) .GT. 2047) rawevent(3) = 0
            if ( rawevent(3) .LT.    0) rawevent(3) = 0
            rawevent(3) = rawevent(3) + ((tel-4)*2048)
            rawevent(4) = ((rawevent(4) - ((tel-4)*2048) + randval2) * final_egain(tel)) + shifte(tel) + 0.5
            if ( rawevent(4) .GT. 2047) rawevent(4) = 0
            if ( rawevent(4) .LT.    0) rawevent(4) = 0
            rawevent(4) = rawevent(4) + ((tel-4)*2048)
         end if                            
C  ***
C  ***   
C  ***

         if (rawevent(3).LT.0.OR.rawevent(3).GT.8192) rawevent(3) = 0
         if (rawevent(4).LT.0.OR.rawevent(4).GT.8192) rawevent(4) = 0

C --     Store E-counter gain-stabilized event in "good" Buffer
         do k = 0, evlen - 1
            OutBuf(OP) = rawevent(k)
            OP = OP + 1
         end do
      end if                             ! End of good event block


C --  Fetch next event
      end do                             ! End-Of-Event Loop
C=======================================================================

C --  Buffer finished or error has occured, return to main
 999  continue
      return
      end                                ! End-Of-Subroutine reduc_winformat

File: sirius/reduc/src/reduc_winformat.c
#include	<stdio.h>
#include	<stdlib.h>
#include	<ctype.h>
#include	<string.h>    
#include	<math.h>
#include	<gsdims.h>

#define boe(x)	((x&0xf0000000)>>28)
#define ndw(x)	(x&0x000000ff)
#define box(x)	((x&0x3f800000)>>23)
#define chn(x)	((x&0x007f0000)>>16)
#define dta(x)	(x&0x0000ffff)
#define EOB     0xe0000000

/* ======================================================== */
/* Purpose :Sort data into a matrix which is used by		*/
/*			the sorting application routine.				*/
/*			The event matrix is processed by the subr.		*/
/*			reduc_winsort which returns whether the event	*/
/*			is to be stored or not.							*/
/* ======================================================== */

int reduc_winsort(int *pesp, int *pdesp, int *pedesp);
int grec, evlen;

int reduc_winformat(int *bufp, int *outp, int *messp, int *pesp, int *pdesp, int *pedesp, float final_edeshift[64], int *pwritestat, int *perrstat,  int newoutlength)
{
	float r;
	float evfrac;
	int ie, i, iE_counter;
	int header_id, header_mask;
	int tpu, nextpointer, temp, xrow, telescopeno;
	int header, xindex, yindex, currentpointer, pattern;
	int good, writestatus;
	int gevents, events, evlen;
	int rawevent[1024];
	int bufferlength; 
	int tel;
	int writestat, errstat;
	int outlength;
	
	int	l_word;
	int	words, nword, nboe, nbox;
	
	writestat = *pwritestat;
	errstat   = *perrstat;
	outlength = newoutlength;
	
	r	= drand48() - 0.5;										/* Random between -0.5 and +0.5	*/
	
	bufferlength	= 32768;
	errstat			= 0;										/*	Initialisering */
	words			= 0;
	
	errstat			= 0;										/*	Initialisering */
	currentpointer  = 0;
	nextpointer		= 0;
	gevents         = 0;
	events          = 0;
	writestat       = 0;
	good			=-1;

	telescopeno		= *(messp + 11);							/* Number of telescopes used */
 
	while (words < bufferlength) {								/* Go through buffer */
		dpnu		= -1;											/* Reset numbers */
		pnu			= -1;             
		nanu		= -1;
		tnanu		= -1;
		nimnu		= -1;
		cadcnu		= -1;
		ctdcnu		= -1;
		punu		= -1;
		scnu		= -1;
        iE_counter	= -1;
		
/*	**************************************************************	*/
/*	Testing header word of event									*/
/*	No more data in buffer if header = 0, return from function		*/
/*	If this should occur too early (partly filled buffer), it may	*/
/*	be a serious problem. Mark such a buffer as bad buffer.			*/
/*	**************************************************************	*/
		currentpointer = nextpointer;
		l_word	= *(bufp + nextpointer);
		nboe	= boe(l_word);
		if(nboe == 0xc){										/* Begin of event */
			nword = ndw(l_word);								/* Number of words (ADCs) with data */
		}else{
			if(l_word != EOB ) errstat = -1;                    /* End of buffer */
			goto L999;											/* Drop current buffer */
		}

/*	Save pointer to next event */
		evlen = nword + 1;
		if( evlen > 1024){
			printf("Warning, eventlength > 1024 \n");
			errstat = -1;
			goto L999;
		}
		nextpointer = evlen + currentpointer;
		
/*	Save the raw event */
		for( i = 0; i < evlen; i++){
			rawevent[i] = *(bufp + currentpointer + i);
		}

/* **************************************** */
/*	Looping through all words (ADCs) and	*/
/*	storing in respective vectors			*/
/* **************************************** */
		for (i = 1; i <= nword; i++){
			l_word	= *(bufp + i + words);
			nbox	= box(l_word);					/* The box that has fired */
			
			if(nbox == 0x00){						/* Pattern ch 0 - 3, NIM ADCs ch 4 - 15*/
				nimnu++;                        
				nimi[nimnu]		=	chn(l_word);
				nim[nimnu]		=	dta(l_word);
			}
			if(nbox == 0x01){						
				nimnu++;                        	/* Wall-clock time  ch 16 - 17*/
				nimi[nimnu]		=	chn(l_word);
				nim[nimnu]		=	dta(l_word);
			}
			if(nbox == 0x02){						/* VME scaler 1151N ch 0-31*/
				scnu++;
				sci[scnu]		=	chn(l_word);
				sc[scnu]		=	dta(l_word);
			}
			if(nbox == 0x10){						/* Time of NaI ch 0-31*/
				tnanu++;
				tnai[tnanu]		=	chn(l_word);
				tna[tnanu]		=	dta(l_word);
			}		
			if(nbox == 0x20){						/* Energy of NaI ch 0-31*/
				nanu++;
				nai[nanu]		=	chn(l_word);
				na[nanu]		=	dta(l_word);
			}
			if(nbox == 0x21){						/* Energy E ch 0-32		*/
				pnu++;								/* Only 8 E counters in */
				pi[pnu]			=	chn(l_word);	/* ch 0, 2,..., 16		*/
				p[pnu]			=	dta(l_word);
			}
			if(nbox == 0x22){						/* Energy dE1 ch 0-31 */
				dpnu++;
				dpi[dpnu]		=	chn(l_word);
				dp[dpnu]		=	dta(l_word);
			}
			if(nbox == 0x23){						/* Energy dE2 ch 32-61 */
				dpnu++;
				dpi[dpnu]		=	chn(l_word) + 32;
				dp[dpnu]		=	dta(l_word);
			}


			
			
			
			
			
#if 0
			if(nbox == 6){						/* Camac ADC1 ch 0-7*/
				cadcnu++;
				cadci[cadcnu]	=	chn(l_word);
				cadc[cadcnu]	=	dta(l_word);
			}
			if(nbox == 7){						/* Camac ADC2 ch 0-7*/
				cadcnu++;
				cadci[cadcnu]	=	chn(l_word) + 8;
				cadc[cadcnu]	=	dta(l_word);
			}
			if(nbox == 8){						/* Camac ADC3 ch 0-7*/
				cadcnu++;
				cadci[cadcnu]	=	chn(l_word) + 16;
				cadc[cadcnu]	=	dta(l_word);
			}
			if(nbox == 9){						/* Camac ADC4 ch 0-7*/
				cadcnu++;
				cadci[cadcnu]	=	chn(l_word) + 24;
				cadc[cadcnu]	=	dta(l_word);
			}
			if(nbox == 10){						/* Camac TDC1 ch 0-7*/
				ctdcnu++;
				ctdci[ctdcnu]	=	chn(l_word);
				ctdc[ctdcnu]	=	dta(l_word) ;
			}
			if(nbox == 11){						/* Camac TDC2 ch 0-7*/
				ctdcnu++;
				ctdci[ctdcnu]	=	chn(l_word) + 8;
				ctdc[ctdcnu]	=	dta(l_word);
			}
			if(nbox == 12){						/* Camac TDC3 ch 0-7*/
				ctdcnu++;
				ctdci[ctdcnu]	=	chn(l_word) + 16;
				ctdc[ctdcnu]	=	dta(l_word);
			}
			if(nbox == 13){						/* Camac TDC4 ch 0-7*/
				ctdcnu++;
				ctdci[ctdcnu]	=	chn(l_word) + 24;
				ctdc[ctdcnu]	=	dta(l_word);
			}
			if(nbox == 14){						/* Camac Pile-up ch 0-1*/
				punu++;
				pui[punu]		=	chn(l_word);
				pu[punu]		=	dta(l_word);
			}
#endif
		}
		words = words + evlen;
		
/*	======================================================	*/
/*  Now, the event is formatted and ready to be sorted		*/
/*	======================================================	*/
		good = reduc_winsort( pesp, pdesp, pedesp);					/* Gate on particles, returns 0 of OK */
		++events;
		if (good == 0 ){											/* Good event, save it */
			++gevents;												/* No of good events */
			evfrac = (gevents * 100) / events;
			*(messp + 14) = (int) evfrac;							/* Percent events accepted */
			if (outlength > bufferlength - evlen) { 
				writestatus = write_exa(outp, messp);				/* Write databuffer to target device */
				if (writestatus == -1){								/* Error in write, terminate session */
					writestat = -1;
					printf("Error in write, terminate session = -1\n");
					goto L999;
				} 
				++grec;												/* Increment "good" buffers */
				*(messp + 15) = grec;								/* Number of accepted buffers */
				outlength = 0;										/* Reset local output buffer pointer */
				for (i = 0; i < bufferlength; ++i){					/* Zeroing local output buffer */
					*(outp + i) = 0;
				}
			}
		}
		
/* ********************************************************************	*/
/* Shift the E detector with the calculated shift. The event has one	*/
/* and only one particle event word for E, located in iE_counter			*/
/* ******************************************************************** */
		p[pnu] = (int)((float)p[pnu] + final_edeshift[pi[pnu]] + r + 0.5);
		if(p[pnu] < 0 | p[pnu] > 2047) p[pnu] = 0;
		
/* ********************************************* */
/* Replace the event word for E with new energy	 */
/* First we zero the energy, then add the energy */
/* ********************************************* */
        rawevent[iE_counter] = rawevent[iE_counter] & 0xffff0000;
        rawevent[iE_counter] = rawevent[iE_counter] + p[pnu];

/* **************************************************** */
/* Store E-counter gain-stabilized event in good Buffer */
/* **************************************************** */
		for (i = 0; i < evlen; ++i){					
            *(outp + outlength + i) = rawevent[i];
		}
		outlength = outlength + evlen;
	}													/* End of good event block */
														/* End-Of-Event Loop, fetch next event */
	L999:
	return outlength;
}
